### ISSUE TITLE
Modify graph metadata using UI

### ISSUE THREAD

**Issue by jlaw9** [Murali-group/GraphSpace#317]
The main missing UI features to modify graph metadata are:
- [ ] Modify graph title
- [ ] Add/remove tags
- [ ] Modify Graph Information / Legend

**Comment by bruce-wayne99** [comment]
I would like to work on this issue.

**Comment by tmmurali** [comment]
@jlaw9 @adbharadwaj Can you help @bruce-wayne99 to get started on fixing this issue?

Thanks for the offer, @bruce-wayne99. We welcome new contributors!

**Comment by adbharadwaj** [comment]
@bruce-wayne99,  Thanks for offering to work on this issue. 

### Step 1
You can find detailed documentation on how the graph information, tags, title, and legend is populated using the following link.

[Graph Data Attributes Attributes Treated Specially by GraphSpace](http://manual.graphspace.org/en/latest/GraphSpace_Network_Model.html#graph-data-attributes-attributes-treated-specially-by-graphspace)

Although the above-mentioned sub-section should be sufficient, I highly encourage you to understand go through entire section on CYJS format to get a more thorough understanding.

### Step 2

Go through our update/put `/graphs/{id}`  API works. Play around with it on your **local setup**.  Try to update all of three fields mentioned above and understand the input and output data format.

Here are some useful links:
1. [API Reference](http://manual.graphspace.org/en/latest/Programmers_Guide.html#api-reference)
2. [Tutorial: How to use GraphSpace APIs](https://summerofcode17.wordpress.com/2017/05/30/using-the-graphspace-restful-api/)

### Step 3

Since the API call for this feature will be from UI, we don't have to authenticate our calls using BasicAuth. Instead, we need to authenticate it using CSRF tokens. 

The good thing is we already have a [JS API](https://github.com/Murali-group/GraphSpace/blob/master/static/js/graphs_page.js#L20) to handle graph update in our [JS library](https://github.com/Murali-group/GraphSpace/blob/master/static/js/graphs_page.js). All you need to worry about is sending in the right input data and hence, step 1 and 2 is really important.

Here is a [sample call](
https://github.com/Murali-group/GraphSpace/blob/master/static/js/graphs_page.js#L691) where we are updating the graph to make it public. 

```js
apis.graphs.update(graph_id, {
                'is_public': 1
            },
            successCallback = function (response) {
                // This method is called when group_to_graph relationship is successfully deleted.
                // The entry from the table is deleted.
                $(e).parent().find('.unshare-graph-btn').removeClass('hidden');
                $(e).addClass('hidden');

            },
            errorCallback = function (xhr, status, errorThrown) {
                // This method is called when  error occurs while deleting group_to_graph relationship.
                $.notify({
                    message: xhr.responseJSON.error_message
                }, {
                    type: 'danger'
                });

            });
```

### Step 4

Write the code !! 

@bruce-wayne99, I would be happy to discuss the user experience flow with you on this thread, if you need to.  Some tips:

- We would prefer to use AJAX request rather than submit and reloading the entire page. 
- Inline editing would be amazing. But, you create a new page for updating graph information.

**Comment by bruce-wayne99** [comment]
@adbharadwaj I actually thought of adding buttons like 'Edit name', 'Add Tag', 'Remove Tag', 'Edit description' in the graph table. As shown in the figure below.
![present-UI](https://user-images.githubusercontent.com/25129274/35159587-ad0c3ce0-fd60-11e7-9c7c-d140db5143a6.png)
 Correct me if there is a better way to provide UI to modify graph meta data.

**Comment by tmmurali** [comment]
@bruce-wayne99 In principle, this UI is good but a challenge is that as we expand the list of items we want to be editable the number of buttons increases. 

Another option to consider is show an "Edit" icon (usually looks like a pen). Upon clicking this icon, either (i) a new page opens up that allows editing of all metadata associated with the graph or 
(ii) the appropriate columns in the table become editable in place.

I like option (i) more since it allows metadata that is not shown in the table to edited. But a minor drawback is that it requires opening a new page.

@adbharadwaj Comments?

**Comment by bruce-wayne99** [comment]
@tmmurali So, Iam thinking of implementing an **Edit** icon for each graph and on clicking the Edit option a new modal opens up just like the way for  **Remove** where for each meta data field an input tag can be specified where he can enter new values for the data and clicking the submit button triggers an AJAX request.
Suggest me if this could be done in a better way.

**Comment by adbharadwaj** [comment]
I like what @tmmurali  suggested in option (i). 

Another option would be to create a new page for editing and that means you will have to add a new url-route. I would rather create a client-side util on the graph page. 

So instead of editing from the graphs list page, users should be editing on the page for a given graph. Something like edit description feature in GitHub.

![Edit description](https://i.imgur.com/CexeWBQ.gif)

Let me know if you have more questions.

**Comment by bruce-wayne99** [comment]
@adbharadwaj i have implemented editing **name** and **title** of the graph.  The following is a gif of the implementation. Please have a look at it.
![updating_name_title](https://user-images.githubusercontent.com/25129274/35440408-46c7494c-02c5-11e8-8af2-549055a4ae6c.gif)
But the page must be reloaded to view the updated table.
I though of two options for handling this scenario:
(i) This could be handled by sending an ajax 'GET' request to render the updated page in the successCallback of apis.graphs.update()
(ii) Using a bootstrap table in the same way for displaying the list of graphs under a user. (has a refresh button which can be clicked to see the updated values).
  Suggest me if there is a better way to handle this scenario or which option should I implement.

**Comment by adbharadwaj** [comment]
@bruce-wayne99 Looks promising!
Can you make it look like Github? You can refer to the GIF in my previous comment. Ideally we wouldn't wanna hide the graph while editing the name and description.  For tags, you can refer to Github's tags edit functionality. Let me know if you have more questions.

**Comment by bruce-wayne99** [comment]
@adbharadwaj since the graph visualization and graph information are on different panels, should i display the name, description, tags, title for the graph in the visualization tab and then provide edit/cancel options for the user in the visualization tab?

**Comment by adbharadwaj** [comment]
@bruce-wayne99 We need to split this into two parts.

1. On graph visualization page, we should allow a user to edit the name, title, and tags of the graph.
2. On graph information page, we should allow a user to edit all of the [graph data attributes](http://manual.graphspace.org/en/latest/GraphSpace_Network_Model.html#graph-data-attributes) like name, title, tags, description etc.

**Comment by bruce-wayne99** [comment]
@adbharadwaj I have implemented editing name,title,description functionality.  gif of the implementation : 
![final](https://user-images.githubusercontent.com/25129274/36088891-1e900b70-0fff-11e8-873c-ef7137b244ec.gif)
I have gone through the tags implementation github but there are using a new page for displaying all the tags and editing them as shown in the screen shot.
![screenshot from 2018-02-12 14-10-04](https://user-images.githubusercontent.com/25129274/36088898-259f2c5c-0fff-11e8-80ea-7c44dcfc60e9.png)
Please suggest me regarding  implementation of add/remove tags.

**Comment by tmmurali** [comment]
Thanks. This implementation looks nice.

Is the edit button seen only by the user of the graph?


On Mon, Feb 12, 2018 at 3:46 AM, Subramanyam Makam <notifications@github.com

**Comment by tmmurali** [comment]
Sorry. I meant to say "owner of the graph".

**Comment by bruce-wayne99** [comment]
Yes, the edit button will only appear if user-id is same as the owner-id.

**Comment by adbharadwaj** [comment]
@bruce-wayne99 Looks great. Few more suggestions. 

- The edit panel should be spread across the entire width of the screen i.e width 100%, instead of centering them.
- Change color of cancel button to red - you can use `danger` class in bootstrap
- To edit tags, you refer to [Manage Topics](https://help.github.com/articles/classifying-your-repository-with-topics/) feature on GitHub. You can refer to the multiple terms input functionality used on graphspace search page to get ideas on how allow users to input multiple tags for graphs.

Let me know if you have any questions.

**Comment by adbharadwaj** [comment]
This is how add topics look like on GitHub.
<img width="1109" alt="screen shot 2018-02-13 at 4 14 36 pm" src="https://user-images.githubusercontent.com/3481095/36174062-126c4130-10d9-11e8-909c-c84f9d67d9a6.png">

**Comment by bruce-wayne99** [comment]
@adbharadwaj Sir, @tmmurali  Sir,
- I have spread the edit panel 100% across the screen.
- Added red color to cancel button (shown in the gif for editing tags).
- Implemented editing/adding tags same as Manage Topics on github
- GIF of the implementation.
![tags](https://user-images.githubusercontent.com/25129274/36279400-06ce45d6-12bd-11e8-8fcf-64ad031e41fb.gif)
- Please suggest me if any changes required. :) :)

**Comment by adbharadwaj** [comment]
@bruce-wayne99 A few more suggestions,

- Font size of Manage tags link looks inconsistent with tags font size left to it.
- Arrange the input field and two buttons in such a way that buttons aligned on the right side and field is bigger size s.t there are no awkward big empty area between field and buttons.
- Edit button is too close to export button. distance should be same as with share button.

Basically, use your discretion to come up with a neat UI.

**Comment by bruce-wayne99** [comment]
@adbharadwaj Sir, 
- I have used to `bootstrap hyperlink` to display `Manage tags`.
- I have aligned `Save` and `Cancel` buttons to the right and spread the `input element` across the whole edit panel in case of both **Add/Remove Tags** and **Edit Name/Description/Title**.
- I have aligned `Edit` button properly with equal spaces between all the buttons.
- GIF of the impelementation:
![ui](https://user-images.githubusercontent.com/25129274/36340608-f8786926-1406-11e8-95c2-0a6f41f41dec.gif)
- Please suggest me if any changes required. :) :)

**Comment by tmmurali** [comment]
@bruce-wayne99 @adbharadwaj This implementation looks good. Two points:

1. Having both an "Edit" button and a "Manage tags" button is bound to be confusing. The user will wonder (before clicking "Edit") about what it does. Can we not unite them in one button?

2. Are we changing the database to record these edits? Will other users who can see the graph get a notification of the changes and what the changes are? Is the plan to add new issues to address these must-have functions? In that case, we should create a separate branch for edit functionality, should we not?

**Comment by bruce-wayne99** [comment]
@adbharadwaj , @tmmurali Sir, 
-  Yes, there might be confusion for the user. So, I will unite them into one button .
-  Yes, I am changing the database to record the edits made  but I have not considered regarding sending the notifications.
- I would like to keep a pull request for the work I have done till now so that we can think about further extending it.

**Comment by adbharadwaj** [comment]
@tmmurali 
1. I agree that we should have one edit button.
2. Yes, the notifications will be automatically generated once the notification framework is merged. IN the notification framework, we currently have a wrapper function on update data access layer function which will trigger the notification, irrespective which element on the UI or API triggers the update.

@bruce-wayne99   Do not worry about the notification in this pull request. That will be taken care when we merge the notifications pull request. For now, just unite the edits under one button and let us know.

**Comment by bruce-wayne99** [comment]
@adbharadwaj, @tmmurali  Sir,
 I have made a pull request for this issue please look at it
